### DISCUSSION TITLE
Constrain a single image or canvas view to a region

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-canvas-panel#5] [How Might We]
I want to render a part of an image (`xywh` region).

This might be part of a Canvas, or a region of an image service (see https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/1#discussioncomment-1129785).

The provision of the target is the same as in the comment linked above: if it's a content state it handles it itself. If it's a canvas or image service, we need to specify a region to constrain to.

If rendering as a simple static or responsive image, we don't need additional behaviour (?).
But if rendering as deep zoom, that implies you can zoom within the viewport provided; but we might want to prevent you panning outside the initial region (or might allow it, as required).
Similarly what happens if a full-screen mode is available and you activate it? What fills the screen and what are the constraints on panning?

**Comment by stephenwf** [comment]
> But if rendering as deep zoom, that implies you can zoom within the viewport provided; but we might want to prevent you panning outside the initial region
There are 2 ways this could be rendered, either as a hard-crop where beyond selected the region is hidden, or constrained where you can still see the image outside of the region. I imagine there may be a use-case for both.

The technical difference in the viewer is either creating a distinct, smaller element and putting it into the viewer or constraining the viewport movement.

**Comment by stephenwf** [comment]
In terms of Atlas, it has a concept of a "zone". A zone is like a viewable area inside of the world. It might be a single canvas, multiple canvases or a single area of a canvas. Original use-cases in mind were a 2-up book view where you paint all of the pages of your book side by side, with a zone for each spread (or 2-up view). That way you can navigate through each zone. 

This is more aligned with the original view of Atlas as a way to blur the line between viewing a full manifest and individual canvases. 

Let me see if I can put together a demo of using a zone with the viewport constraints (this is not supported in the React version)

**Reply by stephenwf** [reply]
The original "zones" API does not support cropping to individual regions of a single image, instead only whole objects (like the 2-up pages example).

**Reply by stephenwf** [reply]
Demo with Atlas zones: #38

**Reply by tomcrane** [reply]
CP doesn't have to know what ranges are. An app that uses CP - e.g., <manifest-strip>, understands a range, and if it wants to draw the range, it uses CP + RegionHighlight to do so. 

Another example - newspaper article on several columns on different pages (3 targets)

Containing app can draw 3 CPs each confined to just that region so you can see just that article parts and not the rest of the pages the parts are on.

https://tomcrane.github.io/iiif-collector/#objects/short-article.json

**Comment by tomcrane** [comment]
Proposal

Extends the img-service and canvas-panel tags from #1 (comment)

```html
// best-fit this region of the source image service into whatever the current viewport size is
<img-service src="http://..." region="900,900,1000,1000" />
<img-service src="http://..." region="pct:25,25,50,50" />

// template developer has a canvas id within a manifest
<canvas-panel
   iiif-content="https://..canvas_id.."
   partof="https://..manifest_id.."
   region="pct:25,25,50,50" 
</canvas-panel>

// The canvas-in-a-manifest might be wholly supplied by a content state:
<canvas-panel
   iiif-content="https://..content_state.."
   region="pct:25,25,50,50" 
</canvas-panel>

// And if that content state is a region of the canvas, then it's all you would need:
<canvas-panel
   iiif-content="https://..content_state.."
</canvas-panel>
```

For all render modes, if the `target` of that content state was a region of canvas (e.g., `..#xywh=900,900,1000,1000`), then it will initialise with that as a best-fit.

```html
// programmatically (applies to img-service too)

<canvas-panel id="cp"></canvas-panel>
<script>
   const vault = new Vault();
   const cp = document.getElementById("cp");
   await vault.loadManifest("..manifest_id..");
   cp.setCanvas("..canvas_id..");
   cp.setAttribute("region", "900,900,1000,1000");

   // you can also move the viewport later, e.g., in a narrative view:
   // ...time passes, handle a user action:   
   cp.setAttribute("region", "2000,1200,456,987");

  // remember that CP is always one canvas.
  // It might be a synthetic canvas that the developer made and composited 
  // other canvases and content onto, but it's always one canvas. That canvas is 
  // accessible via the vault.

  // although a shortcut?
  // let cv = cp.canvas; // equivalent of obtaining it from the vault?

  // Your app might be using a wrapping layout component, e.g., to layout a 
  // manifest as a strip (with zones) so you don't have to explicitly do the
  // compositing, but that tag is probably not called canvas-panel even if it really
  // is canvas panel with an extra synthetic canvas compositing wrapper.

  
</script>   
```

As well as setting region via an attribute, you can also set it in code. The `region` attribute is a 2D-specific simplification of a more general `Target` helper class (see https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/33#discussion-3509526). Here we are creating a Target instance ourselves to specify a part of the canvas.

```js

const tgt = new Target();
tgt.spatial.x = 2000;
tgt.spatial.y = 1200;
tgt.spatial.w = 456;
tgt.spatial.h = 987;

// also a useful helper constructor, when you have it in this form:

const tgt = new Target("xywh=2000,1200,456,987"); // same thing

cp.goToTarget(tgt);

// In this code approach, we can also specify the behaviour of the transition via options:

cp.goToTarget(tgt, { highlight: true, padding: 40 });
cp.goToTarget(tgt, { transition: "transform 500ms ease-out" });

// In the latter, the syntax of transition is the same as CSS transition
// https://developer.mozilla.org/en-US/docs/Web/CSS/transition

```

**Comment by tomcrane** [comment]
There's an unanswered question above which I think is solved like this:

It doesn't apply in these two cases because the render mode is not pan and zoom:

```html
<canvas-panel iiif-content=".." region="800,800,1000,1000" render="static" />
<canvas-panel iiif-content=".." region="800,800,1000,1000" render="responsive" />
```

But if it's the default pan and zoom, the user can obviously zoom in (otherwise we'd have just made it static or responsive), but can they pan outside of the declared `region`? This needs to be an option:

```html
<canvas-panel iiif-content=".." region="800,800,1000,1000" region-locked="true" />
```

(There is probably a better name for that attribute).
